lib/db/observed.go

```diff
@@ -0,0 +1,180 @@
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	for iter.Next() {
+		keyDev := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		deviceID, err := protocol.DeviceIDFromBytes(keyDev)
+		var bs []byte
+		var od ObservedDevice
+		if err != nil {
+			goto deleteKey
+		}
+		if bs, err = db.Get(iter.Key()); err != nil {
+			goto deleteKey
+		}
+		if err = od.Unmarshal(bs); err != nil {
+			goto deleteKey
+		}
+		res[deviceID] = od
+		continue
+	deleteKey:
+		// Deleting invalid entries is the only possible "repair" measure and
+		// appropriate for the importance of pending entries.  They will come back
+		// soon if still relevant.
+		l.Infof("Invalid pending device entry, deleting from database: %x", iter.Key())
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) AddOrUpdatePendingFolder(id, label string, device protocol.DeviceID) error {
+	key, err := db.keyer.GeneratePendingFolderKey(nil, device[:], []byte(id))
+	if err != nil {
+		return err
+	}
+	of := ObservedFolder{
+		Time:  time.Now().Round(time.Second),
+		Label: label,
+	}
+	bs, err := of.Marshal()
+	if err != nil {
+		return err
+	}
+	return db.Put(key, bs)
+}
+
+// RemovePendingFolderForDevice removes entries for specific folder / device combinations.
+func (db *Lowlevel) RemovePendingFolderForDevice(id string, device protocol.DeviceID) {
+	key, err := db.keyer.GeneratePendingFolderKey(nil, device[:], []byte(id))
+	if err != nil {
+		return
+	}
+	if err := db.Delete(key); err != nil {
+		l.Warnf("Failed to remove pending folder entry: %v", err)
+	}
+}
+
+// RemovePendingFolder removes all entries matching a specific folder ID.
+func (db *Lowlevel) RemovePendingFolder(id string) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+	if err != nil {
+		l.Infof("Could not iterate through pending folder entries: %v", err)
+		return
+	}
+	defer iter.Release()
+	for iter.Next() {
+		if id != string(db.keyer.FolderFromPendingFolderKey(iter.Key())) {
+			continue
+		}
+		if err := db.Delete(iter.Key()); err != nil {
+			l.Warnf("Failed to remove pending folder entry: %v", err)
+		}
+	}
+}
+
+// Consolidated information about a pending folder
+type PendingFolder struct {
+	OfferedBy map[protocol.DeviceID]ObservedFolder `json:"offeredBy"`
+}
+
+func (db *Lowlevel) PendingFolders() (map[string]PendingFolder, error) {
+	return db.PendingFoldersForDevice(protocol.EmptyDeviceID)
+}
+
+// PendingFoldersForDevice enumerates only entries matching the given device ID, unless it
+// is EmptyDeviceID.  Invalid ones are dropped from the database after a warning log
+// message, as a side-effect.
+func (db *Lowlevel) PendingFoldersForDevice(device protocol.DeviceID) (map[string]PendingFolder, error) {
+	var err error
+	prefixKey := []byte{KeyTypePendingFolder}
+	if device != protocol.EmptyDeviceID {
+		prefixKey, err = db.keyer.GeneratePendingFolderKey(nil, device[:], nil)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Right, but it's two more lines because `prefixKey` needs to be declared first for both cases.  What would be "The Syncthing Style" to do it?  My impression of other code parts is that separate declaration and assignment is rather avoided.